modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -0,0 +1,712 @@
+
+/*         TsdfType v00 = vx[0] + t.z*(vx[1] - vx[0]); */
+/*         TsdfType v01 = vx[2] + t.z*(vx[3] - vx[2]); */
+/*         TsdfType v10 = vx[4] + t.z*(vx[5] - vx[4]); */
+/*         TsdfType v11 = vx[6] + t.z*(vx[7] - vx[6]); */
+
+/*         TsdfType v0 = v00 + t.y*(v01 - v00); */
+/*         TsdfType v1 = v10 + t.y*(v11 - v10); */
+
+/*         nv = v0 + t.x*(v1 - v0); */
+
+/*         next[c] = voxelIdx[c]; */
+/*         prev[c] = voxelIdx[c]; */
+/*     } */
+
+/*     return normalize(normal); */
+/* } */
+
+struct RaycastInvoker : ParallelLoopBody
+{
+    RaycastInvoker(Points& _points, Normals& _normals, Affine3f cameraPose, Intr intrinsics,
+                   const HashTSDFVolumeCPU& _volume)
+        : ParallelLoopBody(),
+          points(_points),
+          normals(_normals),
+          volume(_volume),
+          tstep(_volume.truncDist * _volume.raycastStepFactor),
+          cam2vol(volume.pose.inv() * cameraPose),
+          vol2cam(cameraPose.inv() * volume.pose),
+          reproj(intrinsics.makeReprojector())
+    {
+    }
+
+    virtual void operator()(const Range& range) const override
+    {
+        const Point3f cam2volTrans = cam2vol.translation();
+        const Matx33f cam2volRot   = cam2vol.rotation();
+        const Matx33f vol2camRot   = vol2cam.rotation();
+
+        const float blockSize = volume.volumeUnitSize;
+
+        for (int y = range.start; y < range.end; y++)
+        {
+            ptype* ptsRow = points[y];
+            ptype* nrmRow = normals[y];
+
+            for (int x = 0; x < points.cols; x++)
+            {
+                //! Initialize default value
+                Point3f point = nan3, normal = nan3;
+
+                //! Ray origin and direction in the volume coordinate frame
+                Point3f orig    = cam2volTrans;
+                Point3f rayDirV = normalize(Vec3f(cam2volRot * reproj(Point3f(x, y, 1.f))));
+
+                float tmin = rgbd::Odometry::DEFAULT_MIN_DEPTH() / rayDirV.z;
+                float tmax = rgbd::Odometry::DEFAULT_MAX_DEPTH() / rayDirV.z;
+
+                float tprev = tmin;
+                float tcurr = tmin;
+
+                TsdfType prevTsdf = volume.truncDist;
+                cv::Vec3i prevVolumeUnitIdx =
+                    cv::Vec3i(std::numeric_limits<int>::min(), std::numeric_limits<int>::min(),
+                              std::numeric_limits<int>::min());
+                cv::Ptr<TSDFVolumeCPU> currVolumeUnit;
+                while (tcurr < tmax)
+                {
+                    Point3f currRayPos          = orig + tcurr * rayDirV;
+                    cv::Vec3i currVolumeUnitIdx = volume.volumeToVolumeUnitIdx(currRayPos);
+
+                    VolumeUnitMap::const_iterator it = volume.volumeUnits.find(currVolumeUnitIdx);
+
+                    TsdfType currTsdf = prevTsdf;
+                    int currWeight    = 0;
+                    float stepSize    = blockSize;
```

Comment:
Such big step can skip a subvolume. The line between `currRayPos` and `currRayPos + blockSize * rayDirV` can enter new subvolume near its corner and then exit it near the same corner.

Reply:
Setting the step to 0.5 * blockSize fixes all the raycasting glitches.